### PR TITLE
feat: add agent trace logger for llm debugging.

### PR DESCRIPTION
**Pull request by blue-phi**
## Summary

- Problem: Developers lack a granular way to debug LLM agent logic, as standard chat logs hide raw API payloads, tool call sequencing, and intermediate reasoning/thinking blocks.
- To make debugging efficient, this PR introduces a transparent **Agent Trace Logger**.
- What changed: Added a event-based logger that captures llm raw payloads, tool execution lifecycle in JSONL format.
- What did NOT change: The core `pi-embedded` execution pipelines and existing user-facing chat history structures remain untouched.

## Change Type (select all)

- [ ] Bug fix
- [x] Feature
- [ ] Refactor
- [ ] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #
- Related #

## User-visible / Behavior Changes

- New environment variables: `OPENCLAW_AGENT_TRACE` (enable/disable), `OPENCLAW_AGENT_TRACE_VERBOSE` (terminal output), and `OPENCLAW_AGENT_TRACE_FILE` (log path).
- Default behavior: Tracing is disabled by default. When enabled, it writes structured event data to `~/.openclaw/logs/agent-trace.jsonl`.

## Security Impact (required)

- New permissions/capabilities? (`No`)
- Secrets/tokens handling changed? (`No`)
- New/changed network calls? (`No`)
- Command/tool execution surface changed? (`No`)
- Data access scope changed? (`No`)

### Environment

- OS: Ubuntu 22.04
- Runtime/container: Node.js / pnpm
- Model/provider: OpenAI-compatible
- Integration/channel: Gateway

### Steps

1. Set `OPENCLAW_AGENT_TRACE=1` and `OPENCLAW_AGENT_TRACE_VERBOSE=1` in your environment.
2. Run `pnpm gateway:dev`.
3. Interact with an agent via the chat interface.
4. Observe real-time output file specified in `OPENCLAW_AGENT_TRACE_FILE`.

### Expected

- `agent-trace.jsonl` contains structured JSON entries for every agent interaction.

### Actual

- Agent internal reasoning and raw provider payloads are successfully captured and persisted.

## Evidence

Attach at least one:

- [ ] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

## Human Verification (required)
[trace_snippets.txt](https://github.com/user-attachments/files/25997892/trace_snippets.txt)

- Verified scenarios: Standard chat, tool calling flow, and reasoning model (thinking block) capture.
- Edge cases checked: Empty payloads, rapid message bursts, and varying provider protocols (Anthropic/OpenAI).
- What you did **not** verify: Long-running stress tests.

## Review Conversations

- [x] I replied to or resolved every bot review conversation I addressed in this PR.
- [x] I left unresolved only the conversations that still need reviewer or maintainer judgment.

## Compatibility / Migration

- Backward compatible? (`Yes`)
- Config/env changes? (`Yes`)
- Migration needed? (`No`)

## Failure Recovery (if this breaks)

- How to disable/revert this change quickly: Unset `OPENCLAW_AGENT_TRACE` (set to 0 or remove env var).
- Files/config to restore: N/A.
- Known bad symptoms reviewers should watch for: Excessive I/O if `agent-trace.jsonl` grows too large (mitigated by non-blocking file writer).

## Risks and Mitigations

None

### CODE REVIEW

File: src/agents/agent-trace-log.ts [hunk 1]
```diff
@@ -0,0 +1,510 @@
...
+
+      case "message_update": {
+        if (!cfg.verbose || role === "user") {
+          break;
+        }
+        const content = message?.content;
+        let text = "";
+        let reasoning = "";
+        if (typeof content === "string") {
+          text = content;
```
[the remark this answers is not in the document]
Comment (blue-phi) [reply src/agents/agent-trace-log.ts:312]: text_delta/thinking_delta have been read and added to message.content.

File: src/agents/agent-trace-log.ts [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+  sessionId?: string;
+  sessionKey?: string;
+  provider?: string;
+  model?: string;
+  // LLM request fields
+  systemPrompt?: string;
+  messagesCount?: number;
+  toolsCount?: number;
+  payload?: unknown;
+  payloadDigest?: string;
```
[the remark this answers is not in the document]
Comment (blue-phi) [reply src/agents/agent-trace-log.ts]: removed payloadDigest as sugguested.

File: src/agents/agent-trace-log.ts [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+        record({
+          ...base,
+          ts: new Date().toISOString(),
+          type: "message_start",
+        });
+        break;
+
+      }
+
+      case "message_end": {
```
[the remark this answers is not in the document]
Comment (blue-phi) [reply src/agents/agent-trace-log.ts:339]: No, message_update events are persisted to the JSONL trace.

**Comment by steipete** [comment]
Closing this as better suited for [ClawHub](https://clawhub.ai/)/community plugin work after Codex review.

Close as out-of-scope for OpenClaw core: current `main` positions tracing/observability in the plugin ecosystem and already exposes plugin hooks for LLM, tool, compaction, and agent-end lifecycle events.

What I checked:

- **Vision scope:** [`VISION.md`](https://github.com/openclaw/openclaw/blob/main/VISION.md) says optional capability should usually ship as plugins, with a high bar for adding optional features to core. ([`VISION.md:54`](https://github.com/openclaw/openclaw/blob/56fe2aab9c87/VISION.md#L54), [56fe2aab9c87](https://github.com/openclaw/openclaw/commit/56fe2aab9c87))
- **Existing trace plugin path:** Community docs already list the official `@opik/opik-openclaw` plugin for exporting agent traces. Public docs: [`docs/plugins/community.md`](https://docs.openclaw.ai/plugins/community). ([`docs/plugins/community.md:77`](https://docs.openclaw.ai/plugins/community), [56fe2aab9c87](https://github.com/openclaw/openclaw/commit/56fe2aab9c87))
- **Conversation trace hooks:** Public plugin hooks already expose `llm_input`, `llm_output`, and `agent_end`, including prompt/history/output/error data for observability integrations. ([`src/plugins/hook-types.ts:179`](https://github.com/openclaw/openclaw/blob/56fe2aab9c87/src/plugins/hook-types.ts#L179), [56fe2aab9c87](https://github.com/openclaw/openclaw/commit/56fe2aab9c87))
- **Tool and compaction lifecycle hooks:** Public plugin hooks also expose `before_compaction`/`after_compaction` and `before_tool_call`/`after_tool_call`, which cover the lifecycle this PR is trying to log. ([`src/plugins/hook-types.ts:221`](https://github.com/openclaw/openclaw/blob/56fe2aab9c87/src/plugins/hook-types.ts#L221), [56fe2aab9c87](https://github.com/openclaw/openclaw/commit/56fe2aab9c87))
- **Explicit access gate for conversation data:** Plugin config already has an `allowConversationAccess` gate specifically for conversation-tracing hooks, showing this is an intended plugin-level surface. ([`src/config/types.plugins.ts:6`](https://github.com/openclaw/openclaw/blob/56fe2aab9c87/src/config/types.plugins.ts#L6), [56fe2aab9c87](https://github.com/openclaw/openclaw/commit/56fe2aab9c87))

So I’m closing this as a scope-fit item for the plugin/community path rather than keeping it open as an OpenClaw core request.

Review notes: reviewed against [56fe2aab9c87](https://github.com/openclaw/openclaw/commit/56fe2aab9c87596515f2280c8060dc740ea64a88).
